src/platform/datapath_raw_xdp.c

```diff
@@ -957,6 +1057,61 @@ CxPlatDpRawUninitialize(
     }
 }
 
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+CxPlatDpRawOnSocketStateChange(
+    _In_ CXPLAT_SOCKET* Socket,
+    _In_ BOOLEAN IsCreated
+    )
+{
+    XDP_DATAPATH* Xdp = (XDP_DATAPATH*)Socket->Datapath;
+
+    if (Socket->Wildcard) {
+        const XDP_RULE Rule = {
+            .Match = XDP_MATCH_UDP_DST,
+            .Pattern.Port = Socket->LocalAddress.Ipv4.sin_port, // TODO - Check on on byte order
+            .Action = XDP_PROGRAM_ACTION_REDIRECT,
+            .Redirect.TargetType = XDP_REDIRECT_TARGET_TYPE_XSK,
+            .Redirect.Target = NULL,
+        };
+
+        CXPLAT_LIST_ENTRY* Entry;
+        for (Entry = Xdp->Interfaces.Flink; Entry != &Xdp->Interfaces; Entry = Entry->Flink) {
+            XDP_INTERFACE* Interface = CONTAINING_RECORD(Entry, XDP_INTERFACE, Link);
+            if (IsCreated) {
+                CxPlatDpRawInterfaceAddRule(Interface, &Rule);
+            } else {
+                CxPlatDpRawInterfaceRemoveRule(Interface, &Rule);
+            }
+        }
+
+    } else {
+
+        //
+        // TODO - Need to do this (1) for the 4-tuple and (2) only on the right
+        // interface. But for now, copy the listener pattern.
+        //
+
+        const XDP_RULE Rule = {
+            .Match = XDP_MATCH_UDP_DST,
+            .Pattern.Port = Socket->LocalAddress.Ipv4.sin_port, // TODO - Check on on byte order
+            .Action = XDP_PROGRAM_ACTION_REDIRECT,
+            .Redirect.TargetType = XDP_REDIRECT_TARGET_TYPE_XSK,
+            .Redirect.Target = NULL,
+        };
+
+        CXPLAT_LIST_ENTRY* Entry;
+        for (Entry = Xdp->Interfaces.Flink; Entry != &Xdp->Interfaces; Entry = Entry->Flink) {
+            XDP_INTERFACE* Interface = CONTAINING_RECORD(Entry, XDP_INTERFACE, Link);
+            if (IsCreated) {
+                CxPlatDpRawInterfaceAddRule(Interface, &Rule);
+            } else {
+                CxPlatDpRawInterfaceRemoveRule(Interface, &Rule);
+            }
+        }
```

Comment:
nit: you can move this part out of the IF ELSE block to reduce code duplication.

```
        CXPLAT_LIST_ENTRY* Entry;
        for (Entry = Xdp->Interfaces.Flink; Entry != &Xdp->Interfaces; Entry = Entry->Flink) {
            XDP_INTERFACE* Interface = CONTAINING_RECORD(Entry, XDP_INTERFACE, Link);
            if (IsCreated) {
                CxPlatDpRawInterfaceAddRule(Interface, &Rule);
            } else {
                CxPlatDpRawInterfaceRemoveRule(Interface, &Rule);
            }
        }
```

Reply:
Yes, we could, but this is temporary. It will (hopefully soon) be specific to connected sockets.